chall04/rarias-p.py

```diff
@@ -0,0 +1,91 @@
+#!/usr/bin/env python3
+
+import sys
+import fileinput
+
+# Power is a function I found looking for ways to get all the possible 
+# combinations of a group of things, regardless the number of items
+# gotten in each combination.
+# http://edupython.blogspot.com/2016/06/potenciando-conjuntos.html
+
+def power(c):                           
+    if len(c) == 0:						
+        return [[]]
+    r = power(c[:-1])
+    return r + [s + [c[-1]] for s in r]
+
+def listsum(numbers):
+	total = 0
+	for num in numbers:
+		total += num
+	return total
+
+def printer(argv, file, lenths):
+	if len(argv) > 2:
+		print(file + ":", end="\n")
+	if len(lenths) == 0:
+		print("./rarias-p: "+ file +": Not enough space in the given shelves")
+	else:
+		print(min(lenths))
+
+def printError(file, argv):
+	if len(argv) > 2:
+		print(file + ":", end="\n")
+	print("./rarias-p: "+ file +": Can't read file")
+
+def checker(lines):
+	shelves = lines[0].split(' ')
+	i = 0
+	try:
+		for number in shelves:
+			shelves[i] = int(number)
+			i += 1
+		totalbooks = 0
+		for line in lines[1:]:
+			totalbooks += int(line.split(' ')[0])
+		possible = []
+		for bookshelf in power(shelves):
+			if listsum(bookshelf) > totalbooks:
+				possible.append(bookshelf)
+		lenths = []
+		for bookshelf in possible:
+			lenths.append(len(bookshelf))
+		return (lenths)
+	except ValueError:
+		pass			# This error is produces another error (line 79) which
+						# is the one that will take care of the expected output.
+						# I think is a bad solution but is the only one I could
+						# come up with.
+
+def main(argv):
+	if len(argv) > 1:
+		fn = 1
+		for file in argv[1:]:
+			try:
+				f = open(argv[fn], "r")
+				lines = f.readlines()
+				lenths = checker(lines)
+				printer(argv, file, lenths)
+				f.close
+				fn += 1
+			except (FileNotFoundError, PermissionError, TypeError):
+				printError(file, argv)
+						# The type error is carried from the value error on 
+						# line 65, this makes the program to pass a empty 
+						# object which has no lenth, that is the one causing 
+						# the problem here. This only means we are getting a
+						# non numeric value from the file.
+				
+	else:
+		try:
+			lines = []
+			for line in fileinput.input():
+				lines.append(line)
+				pass
+			lenths = checker(lines)
+			printer(argv, "stdin", lenths)
+		except TypeError:

```

Comment:
You should try to catch `EOFError` too for safety incase I `Ctrl-D` your script.